### PR TITLE
Add app `@gear-js/upload-dapp-chain`

### PR DESCRIPTION
**Pull request by Almazatun**
Resolves # .

@reviewer-or-team

### CODE REVIEW

File: apps/deploy/upload-dapp-chain/src/common/guards/tg-access-accounts.guard.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+@Injectable()
+export class TgAccessAccountsGuard implements CanActivate {
+  constructor() {}
+
+  canActivate(context: ExecutionContext): boolean {
+    const ctx = TelegrafExecutionContext.create(context);
+    const { from } = ctx.getContext<Context>();
+
+    const accounts = process.env.TELEGRAM_ACCESS_ACCOUNTS.split(",");
```
Comment (osipov-mit) [inline apps/deploy/upload-dapp-chain/src/common/guards/tg-access-accounts.guard.ts:15]: Let's keep it in database. And pls add functionality to add new account of person who can use this bot

File: apps/deploy/upload-dapp-chain/src/common/hellpers/get-list-repo.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
+import { readFileSync } from "fs";
+import { load } from "js-yaml";
+
+// eslint-disable-next-line consistent-return
+export function getListRepo(): { repo: string }[] {
+  const path = "/../../../src/common/repos/dapps-repo.yaml";
```
Comment (osipov-mit) [inline apps/deploy/upload-dapp-chain/src/common/hellpers/get-list-repo.ts:6]: I'm not sure that it is good way to keep it in `src` maybe it will be better to take path from process.env and keep it somewhere on server.

File: apps/deploy/upload-dapp-chain/src/common/hellpers/get-name-dapp.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+  let indexLastIndex;
+
+  for (let i = 0; i < chars.length; i++) {
+    if (Number(chars[i]) || Number(chars[i]) === 0) {
+      indexLastIndex = i - 1;
+      break;
+    }
+  }
+
+  return fullName.substr(0, indexLastIndex);
```
Comment (osipov-mit) [inline apps/deploy/upload-dapp-chain/src/common/hellpers/get-name-dapp.ts:12]: deprecated method

File: apps/deploy/upload-dapp-chain/src/common/hellpers/get-name-dapp.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
+export function getNameDapp(fullName: string): string {
```
Comment (osipov-mit) [inline apps/deploy/upload-dapp-chain/src/common/hellpers/get-name-dapp.ts:1]: Can you describe what's going on here?

File: apps/deploy/upload-dapp-chain/src/common/hellpers/get-account.ts [hunk 5]
```diff
@@ -0,0 +1,8 @@
+import { GearKeyring } from "@gear-js/api";
```
Comment (osipov-mit) [inline apps/deploy/upload-dapp-chain/src/common/hellpers/get-account.ts:1]: typo in the directory name

File: apps/deploy/upload-dapp-chain/src/dapp-data/dapp-data.service.ts [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+    const payloads = {
+      [DAPP.NFT]: { name: "test", symbol: "test", baseUri: "test", royalties: null },
+      [DAPP.NFT_MARKETPLACE]: { adminId: this.configService.get<string>("nftMarketplace.NFT_MARKETPLACE_ADMIN_ID"),
+        treasuryId: this.configService.get<string>("nftMarketplace.NFT_MARKETPLACE_TREASURY_ID"),
+        treasuryFee: "1" },
+      [DAPP.LOTTERY]: "0x00",
+      [DAPP.DUTCH_AUCTION]: "0x00",
+      [DAPP.ESCROW]: "--",
+      [DAPP.SUPPLY_CHAIN]: "--",
+      default: false,
```
Comment (osipov-mit) [inline apps/deploy/upload-dapp-chain/src/dapp-data/dapp-data.service.ts:200]: Let's take this out of the source code. I see a few ways how to do it.
1. Store payloads in database, and add new payloads for new apps manually (through tgbot).
2. Create `.yaml` files for each of the apps and read store it on the server.

File: apps/deploy/upload-dapp-chain/src/dapp-data/dapp-data.service.ts [hunk 7]
```diff
@@ -0,0 +1,217 @@
...
+    };
+
+    return payloads[name] || payloads.default;
+  }
+
+  private getNameDappsByRelease(release: Release): string[] {
```
Comment (osipov-mit) [inline apps/deploy/upload-dapp-chain/src/dapp-data/dapp-data.service.ts:206]: Is it used somewhere?

File: apps/deploy/upload-dapp-chain/src/dapp-data/dapp-data.service.ts [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+    if (!paylaod) {
+      throw new BadRequestExc("Invalid dapp name");
+    }
+
+    const meta = await getWasmMetadata(metaWasmBuff);
+
+    const gas = await gearApi.program.calculateGas.initUpload(
+      sourceId,
+      optWasmBuff,
+      meta.types,
```
Comment (osipov-mit) [inline apps/deploy/upload-dapp-chain/src/dapp-data/dapp-data.service.ts:141]: Why did you add `meta.types` here? The payload should be here.

File: apps/deploy/upload-dapp-chain/src/tgbot/tgbot.service.ts [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+      } else {
+        res = "Please enter valid repo and dapp name 😡";
+      }
+    } else {
+      res = "Invalid command arguments 😡";
+    }
+
+    ctx.reply(res);
+  }
+}
```
Comment (osipov-mit) [inline apps/deploy/upload-dapp-chain/src/tgbot/tgbot.service.ts:63]: We need functionality to upload all apps with one command

**Review by osipov-mit — CHANGES_REQUESTED** [review]

File: apps/deploy/upload-dapp-chain/src/common/hellpers/get-name-dapp.ts (hunk 4 above)
Comment (Almazatun) [reply apps/deploy/upload-dapp-chain/src/common/hellpers/get-name-dapp.ts:1]:
```js
let str = "nft_marketplace-0.1.0.opt.wasm"

getNameDapp(str) // output ---> "nft_marketplace"
```

File: apps/deploy/upload-dapp-chain/src/common/hellpers/get-list-repo.ts (hunk 2 above)
Comment (Almazatun) [reply apps/deploy/upload-dapp-chain/src/common/hellpers/get-list-repo.ts]: moved to `.env`

File: apps/deploy/upload-dapp-chain/src/common/hellpers/get-name-dapp.ts (hunk 3 above)
Comment (Almazatun) [reply apps/deploy/upload-dapp-chain/src/common/hellpers/get-name-dapp.ts]: updated this part

File: apps/deploy/upload-dapp-chain/src/dapp-data/dapp-data.service.ts (hunk 6 above)
Comment (Almazatun) [reply apps/deploy/upload-dapp-chain/src/dapp-data/dapp-data.service.ts]: moved to `yaml` file

File: apps/deploy/upload-dapp-chain/src/dapp-data/dapp-data.service.ts (hunk 8 above)
Comment (Almazatun) [reply apps/deploy/upload-dapp-chain/src/dapp-data/dapp-data.service.ts]: Fixed

File: apps/deploy/upload-dapp-chain/src/dapp-data/dapp-data.service.ts (hunk 7 above)
Comment (Almazatun) [reply apps/deploy/upload-dapp-chain/src/dapp-data/dapp-data.service.ts]: Deleted unused `private method`

File: apps/deploy/upload-dapp-chain/src/configs/base.config.ts [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+  bot: {
+    TELEGRAM_BOT_TOKEN: process.env.TELEGRAM_BOT_TOKEN || "bot",
+  },
+  workflow: {
+    WORKFLOW_PATH: process.env.WORKFLOW_PATH || "path",
+  },
+  nftMarketplace: {
+    NFT_MARKETPLACE_ADMIN_ID: process.env.NFT_MARKETPLACE_ADMIN_ID || "nftMarketplaceId",
+    NFT_MARKETPLACE_TREASURY_ID: process.env.NFT_MARKETPLACE_TREASURY_ID || "treasuryId",
+  },
```
Comment (osipov-mit) [inline apps/deploy/upload-dapp-chain/src/configs/base.config.ts:19]: This is unnecessary in environment variables

File: apps/deploy/upload-dapp-chain/src/workflow-command/workflow-command.service.ts [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+        value: 0,
+      };
+
+      const tx = this.gearApi.message.send(message, meta);
+
+      await sendTransaction(this.gearApi, account, "MessageEnqueued", ApiKey.MESSAGE);
+
+      await tx.signAndSend(account, ({ events }) => {
+        events.forEach(({ event }) => console.log(event.toHuman()));
+      });
```
Comment (osipov-mit) [inline apps/deploy/upload-dapp-chain/src/workflow-command/workflow-command.service.ts:99]: Why are you sending the transaction 2 times?

File: apps/deploy/upload-dapp-chain/src/workflow-command/workflow-command.service.ts [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+          0,
+          true,
+          meta,
+      );
+
+      const message: IMessageSendOptions = {
+        destination: uploadedProgram.programId as Hex,
+        payload,
+        gasLimit: gas.min_limit,
+        value: 0,
```
Comment (osipov-mit) [inline apps/deploy/upload-dapp-chain/src/workflow-command/workflow-command.service.ts:90]: We need to take value from the workflow file. If value is not specified there, then undefiend is normal

File: apps/deploy/upload-dapp-chain/test/jest-e2e.json [hunk 13]
```diff
@@ -0,0 +1,9 @@
+{
```
Comment (osipov-mit) [inline apps/deploy/upload-dapp-chain/test/jest-e2e.json:1]: Is this file necessary?

**Review by osipov-mit — CHANGES_REQUESTED** [review]

File: apps/deploy/upload-dapp-chain/src/common/workflow/workflow.yaml [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+      acc: "alice"
+      payload: "0x00"
+
+    6:
+      dapp: "escrow"
+      repo: "escrow"
+      optDownloadUrl: ""
+      metaDownloadUrl: ""
+      acc: "alice"
+      payload: ""
```
[the remark this answers is not in the document]
Comment (Almazatun) [reply apps/deploy/upload-dapp-chain/src/common/workflow/workflow.yaml:59]: Deleted the current program

File: apps/deploy/upload-dapp-chain/src/workflow-command/workflow-command.service.ts (hunk 11 above)
Comment (Almazatun) [reply apps/deploy/upload-dapp-chain/src/workflow-command/workflow-command.service.ts:99]: Deleted second `transaction`

File: apps/deploy/upload-dapp-chain/src/configs/base.config.ts (hunk 10 above)
Comment (Almazatun) [reply apps/deploy/upload-dapp-chain/src/configs/base.config.ts:19]: Deleted

File: apps/deploy/upload-dapp-chain/src/common/hellpers/get-tg-commands.ts [hunk 15]
```diff
@@ -0,0 +1,5 @@
+export function getTgCommands(): string {
+  return "/uploadDapps" + "\n"
+  + "/getUserId" + "\n"
+  + "/addAccessUser [telegramUserId]";
+}
```
[the remark this answers is not in the document]
Comment (Almazatun) [reply apps/deploy/upload-dapp-chain/src/common/hellpers/get-tg-commands.ts]: Included

File: apps/deploy/upload-dapp-chain/src/common/hellpers/send-transaction.ts [hunk 16]
```diff
@@ -0,0 +1,27 @@
+import { KeyringPair } from "@polkadot/keyring/types";
+import { GearApi } from "@gear-js/api";
+
+import { ApiKey } from "../enums";
+
+export async function sendTransaction(
+  gearApi: GearApi,
```
[the remark this answers is not in the document]
Comment (osipov-mit) [reply apps/deploy/upload-dapp-chain/src/common/hellpers/send-transaction.ts:7]: Yes

**Review by nikitayutanov — APPROVED** [review]

**Review by osipov-mit — APPROVED** [review]
